CI failure: E2E Tests — Agent creation and SSH timeout failures

## CI Failure Report

**Workflow:** E2E Tests
**Branch:** main
**Commit:** 8a14b27881d0173cc0a1364cf624c9cfee3f129c
**Run:** https://github.com/Action-Llama/action-llama/actions/runs/23412821291

## Failure Summary

The E2E test suite is failing with two main issues: (1) CLI tests cannot find or run agents after creation, and (2) all deployment tests fail due to SSH service timeout in VPS containers.

## Failed Job(s)

### e2e

**Error output:**
```
❯ CLI Flows > creates and runs agent
Error: Error: Agent "test-agent" not found. No agents found.
 ❯ Socket.<anonymous> src/harness.ts:206:18

❯ CLI Flows > manages agent lifecycle
AssertionError: expected 'AL Status — /home/testuser/test-proje…' to contain 'Running'
Expected: Running
Received: AL Status — /home/testuser/test-project
Agents:
AGENT           TRIGGER         STATUS      INSTANCES   QUEUE
----------------------------------------------------------------

❯ Deployment Flows > deploys to VPS (and 3 other tests)
Error: SSH service failed to start within timeout
 ❯ E2ETestContext.waitForSSH src/harness.ts:307:11
 ❯ E2ETestContext.createVPSContainer src/harness.ts:175:5
```

## Root Cause Analysis

### 1. Agent Creation/Discovery Issue
The CLI flow tests are failing because agents appear to not be created or registered properly:
- In `cli-flows.test.ts:75`, after creating "test-agent", the `al run test-agent` command fails with "Agent 'test-agent' not found"
- In the lifecycle test, `al stat` shows an empty agent list, suggesting the scheduler isn't finding any agents

### 2. SSH Service Timeout
All deployment tests fail at the VPS container setup stage:
- The `waitForSSH` method in `harness.ts:307` tries to connect via SSH for 30 attempts (30 seconds) but never succeeds
- This prevents any deployment tests from running their actual test logic

## Suggested Fix

### For Agent Creation Issue:
1. Check if the agent creation in `createTestAgent` is properly writing files to the expected location
2. Verify the project structure expected by `al run` and `al stat` commands
3. Ensure the recent commit (8a14b27) that added `triggerAgent` to IntegrationHarness didn't break compatibility with E2E tests

### For SSH Timeout Issue:
1. Increase the SSH timeout or add better diagnostics to understand why SSH isn't starting
2. Check if the Docker container for VPS simulation is properly configured with SSH service
3. Verify the container networking setup allows SSH connections from the test harness

## Additional Context

The most recent commit (8a14b27) added a `triggerAgent` method to the integration harness but not to the E2E harness. While this shouldn't directly cause these failures, it's worth investigating if there were related changes that might have affected the test environment.

---
*This issue was automatically created by the gh-actions-responder agent.*

--- a/packages/e2e/docker/vps/Dockerfile
+++ b/packages/e2e/docker/vps/Dockerfile
@@ -51,17 +51,56 @@ EXPOSE 22
 RUN echo '#!/bin/bash\n\
 set -e\n\
 \n\
-# Start Docker daemon in background\n\
-dockerd &\n\
+echo "Starting VPS container services..."\n\
 \n\
-# Wait for Docker to be ready\n\
-while ! docker info > /dev/null 2>&1; do\n\
-    echo "Waiting for Docker daemon to start..."\n\
+# Ensure SSH host keys exist\n\
+if [ ! -f /etc/ssh/ssh_host_rsa_key ]; then\n\
+    ssh-keygen -t rsa -f /etc/ssh/ssh_host_rsa_key -N ""\n\
+fi\n\
+if [ ! -f /etc/ssh/ssh_host_ecdsa_key ]; then\n\
+    ssh-keygen -t ecdsa -f /etc/ssh/ssh_host_ecdsa_key -N ""\n\
+fi\n\
+if [ ! -f /etc/ssh/ssh_host_ed25519_key ]; then\n\
+    ssh-keygen -t ed25519 -f /etc/ssh/ssh_host_ed25519_key -N ""\n\
+fi\n\
+\n\
+# Set proper permissions\n\
+chmod 600 /etc/ssh/ssh_host_*_key\n\
+chmod 644 /etc/ssh/ssh_host_*_key.pub\n\
+\n\
+# Ensure SSH directory structure\n\
+mkdir -p /var/run/sshd\n\
+mkdir -p /root/.ssh\n\
+chmod 700 /root/.ssh\n\
+\n\
+# Start Docker daemon in background\n\
+echo "Starting Docker daemon..."\n\
+dockerd > /tmp/dockerd.log 2>&1 &\n\
+DOCKER_PID=$!\n\
+\n\
+# Wait for Docker to be ready (with timeout)\n\
+echo "Waiting for Docker daemon to start..."\n\
+timeout=30\n\
+while [ $timeout -gt 0 ] && ! docker info > /dev/null 2>&1; do\n\
     sleep 1\n\
+    timeout=$((timeout - 1))\n\
 done\n\
 \n\
+if [ $timeout -eq 0 ]; then\n\
+    echo "Docker daemon failed to start within timeout"\n\
+    kill $DOCKER_PID 2>/dev/null || true\n\
+    exit 1\n\
+fi\n\
+\n\
+echo "Docker daemon started successfully"\n\
+\n\
+# Test SSH daemon\n\
+echo "Testing SSH daemon configuration..."\n\
+/usr/sbin/sshd -t\n\
+\n\
 # Start SSH daemon\n\
-exec /usr/sbin/sshd -D\n\
+echo "Starting SSH daemon..."\n\
+exec /usr/sbin/sshd -D -e\n\
 ' > /start.sh && chmod +x /start.sh
 
 CMD ["/start.sh"]

--- a/packages/e2e/src/containers/local.ts
+++ b/packages/e2e/src/containers/local.ts
@@ -66,19 +66,84 @@ ${skill}`;
 ${skillContent}
 EOF`
   ]);
+  
+  // Create agent-config.json if it doesn't exist (some AL versions expect this)
+  await context.executeInContainer(containerInfo, [
+    "bash", "-c", `cat > /home/testuser/test-project/${agentName}/agent-config.json << 'EOF'
+{
+  "name": "${agentName}",
+  "model": "claude-3-5-sonnet-20241022",
+  "schedule": "0 */6 * * *"
+}
+EOF`
+  ]);
+  
+  // Verify the agent was created properly
+  const agentFiles = await context.executeInContainer(containerInfo, [
+    "ls", "-la", `/home/testuser/test-project/${agentName}/`
+  ]);
+  
+  if (!agentFiles.includes("SKILL.md")) {
+    throw new Error(`Failed to create agent ${agentName}: SKILL.md not found`);
+  }
+  
+  // Ensure correct ownership and permissions
+  await context.executeInContainer(containerInfo, [
+    "bash", "-c", `chown -R testuser:testuser /home/testuser/test-project/${agentName}`
+  ]);
 }
 
 export async function startActionLlamaScheduler(
   context: E2ETestContext,
   containerInfo: ContainerInfo
 ): Promise<void> {
+  // Ensure we're in the correct directory and verify project structure
+  const projectCheck = await context.executeInContainer(containerInfo, [
+    "bash", "-c", "cd /home/testuser/test-project && ls -la"
+  ]);
+  
+  if (!projectCheck.includes("project.toml")) {
+    throw new Error("Project configuration not found before starting scheduler");
+  }
+  
   // Start the scheduler in the background
   await context.executeInContainer(containerInfo, [
     "bash", "-c", "cd /home/testuser/test-project && nohup al start > /tmp/scheduler.log 2>&1 & echo $! > /tmp/scheduler.pid"
   ]);
   
-  // Wait a bit for scheduler to start
-  await new Promise(resolve => setTimeout(resolve, 5000));
+  // Wait for scheduler to start and verify it's running
+  let attempts = 0;
+  const maxAttempts = 10;
+  
+  while (attempts < maxAttempts) {
+    await new Promise(resolve => setTimeout(resolve, 1000));
+    
+    try {
+      // Check if the scheduler process is still running
+      const pidCheck = await context.executeInContainer(containerInfo, [
+        "bash", "-c", "if [ -f /tmp/scheduler.pid ]; then ps -p $(cat /tmp/scheduler.pid) > /dev/null && echo 'running' || echo 'not running'; else echo 'no pid file'; fi"
+      ]);
+      
+      if (pidCheck.includes("running")) {
+        // Additional verification that scheduler is responding
+        const statusCheck = await context.executeInContainer(containerInfo, [
+          "bash", "-c", "cd /home/testuser/test-project && al stat 2>&1 || echo 'stat failed'"
+        ]);
+        
+        if (!statusCheck.includes("stat failed")) {
+          return; // Scheduler is running properly
+        }
+      }
+    } catch {
+      // Continue trying
+    }
+    
+    attempts++;
+  }
+  
+  // If we get here, the scheduler didn't start properly
+  const logs = await getSchedulerLogs(context, containerInfo);
+  throw new Error(`Scheduler failed to start properly after ${maxAttempts} attempts. Logs: ${logs}`);
 }
 
 export async function stopActionLlamaScheduler(
@@ -113,7 +178,42 @@ export async function runSingleAgent(
   containerInfo: ContainerInfo,
   agentName: string
 ): Promise<string> {
-  return await context.executeInContainer(containerInfo, [
-    "bash", "-c", `cd /home/testuser/test-project && al run ${agentName}`
-  ]);
+  // First verify the agent exists and the project structure is correct
+  try {
+    const projectContents = await context.executeInContainer(containerInfo, [
+      "bash", "-c", "cd /home/testuser/test-project && find . -name 'SKILL.md' -o -name 'project.toml'"
+    ]);
+    
+    if (!projectContents.includes(`${agentName}/SKILL.md`)) {
+      throw new Error(`Agent ${agentName} not found in project. Found files: ${projectContents}`);
+    }
+    
+    // Check if AL can see the agent
+    const alStatus = await context.executeInContainer(containerInfo, [
+      "bash", "-c", `cd /home/testuser/test-project && al stat || echo "al stat failed"`
+    ]);
+    
+    // Run the agent with explicit error handling
+    return await context.executeInContainer(containerInfo, [
+      "bash", "-c", `cd /home/testuser/test-project && al run ${agentName} 2>&1`
+    ]);
+  } catch (error: any) {
+    // Provide better error context for debugging
+    let debugInfo = "";
+    try {
+      const workingDir = await context.executeInContainer(containerInfo, [
+        "bash", "-c", "pwd && ls -la /home/testuser/test-project"
+      ]);
+      debugInfo += `Working directory: ${workingDir}\n`;
+      
+      const alVersion = await context.executeInContainer(containerInfo, [
+        "bash", "-c", "al --version || echo 'al not found'"
+      ]);
+      debugInfo += `AL version: ${alVersion}\n`;
+    } catch {
+      debugInfo = "Could not gather debug info";
+    }
+    
+    throw new Error(`Failed to run agent ${agentName}: ${error.message}\nDebug info: ${debugInfo}`);
+  }
 }

--- a/packages/e2e/src/harness.ts
+++ b/packages/e2e/src/harness.ts
@@ -284,16 +284,79 @@ export class E2ETestContext {
     });
   }
 
-  private async waitForSSH(containerInfo: ContainerInfo, maxAttempts = 30): Promise<void> {
+  private async waitForSSH(containerInfo: ContainerInfo, maxAttempts = 60): Promise<void> {
+    if (!containerInfo.ipAddress) {
+      throw new Error("Container IP address not available for SSH connection");
+    }
+
     for (let i = 0; i < maxAttempts; i++) {
       try {
+        // First check if SSH port is open with a simple connection test
+        await this.testSSHConnection(containerInfo);
+        
+        // Then verify SSH actually works with a command
         await this.executeSSHCommand(containerInfo, "echo 'SSH Ready'");
         return;
-      } catch {
-        await new Promise(resolve => setTimeout(resolve, 1000));
+      } catch (error: any) {
+        if (i < maxAttempts - 1) {
+          // Wait progressively longer for the first few attempts to allow service startup
+          const delay = i < 10 ? 2000 : 1000;
+          await new Promise(resolve => setTimeout(resolve, delay));
+        } else {
+          // On final attempt, include more diagnostic info
+          try {
+            const containerLogs = await this.getContainerLogs(containerInfo);
+            throw new Error(`SSH service failed to start within timeout. Container logs: ${containerLogs.slice(-1000)}`);
+          } catch {
+            throw new Error(`SSH service failed to start within timeout after ${maxAttempts} attempts. IP: ${containerInfo.ipAddress}`);
+          }
+        }
       }
     }
-    throw new Error("SSH service failed to start within timeout");
+  }
+
+  private async testSSHConnection(containerInfo: ContainerInfo): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const conn = new SSHClient();
+      
+      const timeout = setTimeout(() => {
+        conn.end();
+        reject(new Error("SSH connection timeout"));
+      }, 5000);
+      
+      conn.on("ready", () => {
+        clearTimeout(timeout);
+        conn.end();
+        resolve();
+      });
+      
+      conn.on("error", (err) => {
+        clearTimeout(timeout);
+        reject(err);
+      });
+      
+      conn.connect({
+        host: containerInfo.ipAddress,
+        port: 22,
+        username: "root",
+        privateKey: this.sshKeyPair.privateKey,
+        readyTimeout: 5000,
+      });
+    });
+  }
+
+  private async getContainerLogs(containerInfo: ContainerInfo): Promise<string> {
+    try {
+      const container = this.docker.getContainer(containerInfo.id);
+      const stream = await container.logs({
+        stdout: true,
+        stderr: true,
+        tail: 50
+      });
+      return stream.toString();
+    } catch {
+      return "Could not retrieve container logs";
+    }
   }
 
   getPrivateKeyPath(): string {

